autoload/bm.vim

```diff
@@ -46,8 +46,11 @@ endfunction
 
 function! bm#update_bookmark_for_sign(file, sign_idx, new_line_nr, new_content)
   let bookmark = bm#get_bookmark_by_sign(a:file, a:sign_idx)
-  call bm#del_bookmark_at_line(a:file, bookmark['line_nr'])
+  let old_line_nr = bookmark['line_nr']
   call bm#add_bookmark(a:file, a:sign_idx, a:new_line_nr, a:new_content, bookmark['annotation'])
+  if old_line_nr !=# a:new_line_nr
+    unlet g:line_map[a:file][old_line_nr]
```

Comment:
When looking at the implementation of `bm#del_bookmark_at_line()` I can see more things than just `unlet g:line_map[a:file][old_line_nr]` happening. How can we be sure we don't get into an inconsistent state by not calling `bm#del_bookmark_at_line()`?

Reply:
As the code is currently, it will yield exactly the same state. The only map key that can change is the line number, so there's no need to delete everything first. It is for performance, I suppose, but really I don't like `del` removing data from the maps that gets immediately recreated by `add`. We know the file remains the same, so there's no need to try removing those whole arrays, they're still needed. The sign index is unchanged, so that's okay to overwrite. The line number *may* change, so this lets `add` overwrite it, and only cleans up the previous map entry if it did change.

Reply:
Thanks for your explanation @Roy-Orbison. I re-visited the code and now understand your reasoning. Thanks for the optimisation 🙏